### PR TITLE
BUG: cal attribute of aggregate rules was not being propagated

### PR DESCRIPTION
**Pull request by richafrank**
Otherwise algos crash with
```python
AttributeError: 'BeforeClose' object has no attribute 'cal' 
```
if a function is scheduled with, e.g.
```python
schedule_function(
    func=my_func,
    date_rule=date_rules.every_day(),
    time_rule=time_rules.market_close(minutes=30) & time_rules.every_minute(),
)
```

### CODE REVIEW

**Comment by ssanderson** [comment]
Just to make sure I understand, `time_rules.market_close(minutes=30) & time_rules.every_minute(),` should be functionally equivalent to just `time_rules.market_close(minutes=30)`, right?

File: zipline/algorithm.py [hunk 1]
```diff
@@ -1063,7 +1063,7 @@ def fetch_csv(self,
 
         return csv_data_source
 
-    def add_event(self, rule=None, callback=None):
+    def add_event(self, rule, callback=None):
```
Comment (ssanderson) [inline zipline/algorithm.py]: Is there a reason `callback` is allowed to be None here?  I would think failing to pass a callback is almost always a bug.

File: zipline/utils/events.py [hunk 2]
```diff
@@ -546,11 +557,14 @@ class StatefulRule(EventRule):
     def __init__(self, rule=None):
         self.rule = rule or Always()
 
-    def new_should_trigger(self, callable_):
```
Comment (ssanderson) [inline zipline/utils/events.py]: Why was this removed?

**Comment by richafrank** [comment]
Yup, semantically there's no reason there for the second term; this was just to show that each of those terms doesn't get assigned a `cal` attribute.

File: zipline/utils/events.py [hunk 3]
```diff
@@ -236,6 +236,8 @@ def handle_data(self, context, data, dt):
 
 
 class EventRule(six.with_metaclass(ABCMeta)):
+    cal = None  # A calendar instance is set when scheduling a function.
```
Comment (ssanderson) [inline zipline/utils/events.py]: Maybe worth clarifying that the calendar is set on instances of the rule, as opposed to the class?

Comment (ssanderson) [reply zipline/utils/events.py]: Should we make `cal` an abstract property to ensure that all the subclasses implement setters?

**Comment by richafrank** [comment]
Agreed. I went through the same thought process, and settled on this PR.

File: zipline/utils/events.py (hunk 2 above)
Comment (richafrank) [reply zipline/utils/events.py]: I didn't find any call sites.

File: zipline/algorithm.py (hunk 1 above)
Comment (richafrank) [reply zipline/algorithm.py]: I'm guessing it was convenient for tests. Maybe @llllllllll knows more.

File: zipline/utils/events.py (hunk 3 above)
Comment (richafrank) [reply zipline/utils/events.py]: I thought about that as well. My only concern was about the performance of the properties inside the `should_trigger` implementation, but I didn't validate that, and `@lazyval` could also solve that.

Comment (ssanderson) [reply zipline/utils/events.py]: I think for simple attribute access `property `will actually be faster than `lazyval`.  lazyval has more complicated machinery in exchange for memoizing the result, but if the fetching the attribute is trivial, then I think property will probably have less overhead.

File: zipline/algorithm.py (hunk 1 above)
Comment (llllllllll) [reply zipline/algorithm.py]: Yeah, I don't think there are good defaults for either. We should make them required.
